{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def debruijnize(reads):\n",
    "    nodes = set()\n",
    "    not_starts = set()\n",
    "    edges = []\n",
    "    for r in reads:\n",
    "        r1 = r[:-1]\n",
    "        r2 = r[1:]\n",
    "        nodes.add(r1)\n",
    "        nodes.add(r2)\n",
    "        edges.append((r1, r2))\n",
    "        not_starts.add(r2)\n",
    "    return (nodes, edges, list(nodes - not_starts))\n",
    "\n",
    "\n",
    "def build_k_mer(str, k):\n",
    "    return [str[i : k + i] for i in range(0, len(str) - k + 1)]\n",
    "\n",
    "\n",
    "def make_node_edge_map(edges):\n",
    "    node_edge_map = {}\n",
    "    for e in edges:\n",
    "        n = e[0]\n",
    "        if n in node_edge_map:\n",
    "            node_edge_map[n].append(e[1])\n",
    "        else:\n",
    "            node_edge_map[n] = [e[1]]\n",
    "    return node_edge_map\n",
    "\n",
    "\n",
    "def eulerian_trail(m, v):\n",
    "    nemap = m\n",
    "    result_trail = []\n",
    "    start = v\n",
    "    result_trail.append(start)\n",
    "    while True:\n",
    "        trail = []\n",
    "        previous = start\n",
    "        while True:\n",
    "\n",
    "            if previous not in nemap:\n",
    "                break\n",
    "            next = nemap[previous].pop()\n",
    "            if len(nemap[previous]) == 0:\n",
    "                nemap.pop(previous, None)\n",
    "            trail.append(next)\n",
    "            if next == start:\n",
    "                break\n",
    "            previous = next\n",
    "        # completed one trail\n",
    "        print(trail)\n",
    "        index = result_trail.index(start)\n",
    "        result_trail = (\n",
    "            result_trail[0 : index + 1]\n",
    "            + trail\n",
    "            + result_trail[index + 1 : len(result_trail)]\n",
    "        )\n",
    "        # choose new start\n",
    "        if len(nemap) == 0:\n",
    "            break\n",
    "        found_new_start = False\n",
    "        for n in result_trail:\n",
    "            if n in nemap:\n",
    "                start = n\n",
    "                found_new_start = True\n",
    "                break  # from for loop\n",
    "        if not found_new_start:\n",
    "            print(\"error\")\n",
    "            print(\"result_trail\", result_trail)\n",
    "            print(nemap)\n",
    "            break\n",
    "    return result_trail\n",
    "\n",
    "\n",
    "def visualize_debruijn(G):\n",
    "    nodes = G[0]\n",
    "    edges = G[1]\n",
    "    dot_str = 'digraph \"DeBruijn graph\" {\\n '\n",
    "    for node in nodes:\n",
    "        dot_str += '    %s [label=\"%s\"] ;\\n' % (node, node)\n",
    "    for src, dst in edges:\n",
    "        dot_str += \"    %s->%s;\\n\" % (src, dst)\n",
    "    return dot_str + \"}\\n\"\n",
    "\n",
    "\n",
    "def assemble_trail(trail):\n",
    "    if len(trail) == 0:\n",
    "        return \"\"\n",
    "    result = trail[0][:-1]\n",
    "    for node in trail:\n",
    "        result += node[-1]\n",
    "    return result\n",
    "\n",
    "\n",
    "def test_assembly_debruijn(t, k):\n",
    "    reads = build_k_mer(t, k)\n",
    "    G = debruijnize(reads)\n",
    "    v = visualize_debruijn(G)\n",
    "    nemap = make_node_edge_map(G[1])\n",
    "    print(G)\n",
    "    print(v)\n",
    "    start = next(iter(G[2])) if (len(G[2]) > 0) else next(iter(G[0]))\n",
    "    trail = eulerian_trail(nemap, start)\n",
    "    return assemble_trail(trail)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "sequences must have the same length if coordinates is None",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 109\u001b[0m\n\u001b[1;32m    107\u001b[0m \u001b[38;5;66;03m# Convert sequences into Seq objects for motif analysis.\u001b[39;00m\n\u001b[1;32m    108\u001b[0m seqs \u001b[38;5;241m=\u001b[39m [Seq(sequence) \u001b[38;5;28;01mfor\u001b[39;00m sequence \u001b[38;5;129;01min\u001b[39;00m sequences\u001b[38;5;241m.\u001b[39mvalues()]\n\u001b[0;32m--> 109\u001b[0m m \u001b[38;5;241m=\u001b[39m \u001b[43mmotifs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcreate\u001b[49m\u001b[43m(\u001b[49m\u001b[43mseqs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mataaat\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    110\u001b[0m motifs_found \u001b[38;5;241m=\u001b[39m m\u001b[38;5;241m.\u001b[39mconsensus\n\u001b[1;32m    112\u001b[0m motifs_found\n",
      "File \u001b[0;32m~/Workspaces/Study/bioinformatics/venv/lib/python3.12/site-packages/Bio/motifs/__init__.py:39\u001b[0m, in \u001b[0;36mcreate\u001b[0;34m(instances, alphabet)\u001b[0m\n\u001b[1;32m     37\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcreate\u001b[39m(instances, alphabet\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mACGT\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m     38\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Create a Motif object.\"\"\"\u001b[39;00m\n\u001b[0;32m---> 39\u001b[0m     alignment \u001b[38;5;241m=\u001b[39m \u001b[43mAlignment\u001b[49m\u001b[43m(\u001b[49m\u001b[43minstances\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     40\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m Motif(alignment\u001b[38;5;241m=\u001b[39malignment, alphabet\u001b[38;5;241m=\u001b[39malphabet)\n",
      "File \u001b[0;32m~/Workspaces/Study/bioinformatics/venv/lib/python3.12/site-packages/Bio/Align/__init__.py:1139\u001b[0m, in \u001b[0;36mAlignment.__init__\u001b[0;34m(self, sequences, coordinates)\u001b[0m\n\u001b[1;32m   1137\u001b[0m             coordinates \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39marray([[\u001b[38;5;241m0\u001b[39m, length]] \u001b[38;5;241m*\u001b[39m \u001b[38;5;28mlen\u001b[39m(sequences))\n\u001b[1;32m   1138\u001b[0m         \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1139\u001b[0m             \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   1140\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msequences must have the same length if coordinates is None\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1141\u001b[0m             )\n\u001b[1;32m   1142\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcoordinates \u001b[38;5;241m=\u001b[39m coordinates\n",
      "\u001b[0;31mValueError\u001b[0m: sequences must have the same length if coordinates is None"
     ]
    }
   ],
   "source": [
    "# Parsing the file content and identifying potential motifs across all sequences.\n",
    "\n",
    "sequences = {\n",
    "    \"test1\": \"acaaccatatatagtagccactgaat\",\n",
    "    \"test2\": \"ccaccccatatatagtagtgcgggtggtg\",\n",
    "    \"test3\": \"ccataaatagataggcagactgtcgctgt\",\n",
    "    \"test4\": \"gtaaacataccataaatagga\",\n",
    "    \"test5\": \"ttcaagaaactgccataaatagcgat\",\n",
    "    \"test6\": \"tagaggtttttgtgccataaataggt\",\n",
    "    \"test7\": \"ccccataaataggaatatcgggatga\",\n",
    "    \"test8\": \"ttgccattaatagattataccatatatgg\",\n",
    "    \"test9\": \"tatcaacaacgataccaacccatatatgg\",\n",
    "    \"test10\": \"tttccaaatatagaaggtgtggaaag\",\n",
    "    \"test11\": \"tccaaatatagtaaaatcgagtcgcggat\",\n",
    "    \"test12\": \"gactggggcccaaatatagcatgttc\",\n",
    "    \"test13\": \"atcattagcttttacttaccataaatgg\",\n",
    "    \"test14\": \"attcttttgccataaatggtaactcg\",\n",
    "    \"test15\": \"ccataaatggcaagtctgtcgaataacgg\",\n",
    "    \"test16\": \"cccataaatggcagggtattagcacg\",\n",
    "    \"test17\": \"ccaaaaatagatatgtgtcgtaacagctt\",\n",
    "    \"test18\": \"ccaaaaatagggggacaatggaagtgggg\",\n",
    "    \"test19\": \"ccaaaaataggccagacgtgtttacaacg\",\n",
    "    \"test20\": \"ccaaaaatagttaaataatgtcatacatt\",\n",
    "    \"test21\": \"ctacaccttccaaaaatagtaatct\",\n",
    "    \"test22\": \"ttgccaaatatggggttagagtgttc\",\n",
    "    \"test23\": \"gtctttaccaaaaatggtgatcctgt\",\n",
    "    \"test24\": \"ttgccaaaaatggagcgtttaccaat\",\n",
    "    \"test25\": \"atccaccatttatagaaagttcaggaggc\",\n",
    "    \"test26\": \"gcataagagaacattccatttatagg\",\n",
    "    \"test27\": \"tcaaccccatttatagccacgtcagt\",\n",
    "    \"test28\": \"catccattaatagtagccataatggcg\",\n",
    "    \"test29\": \"ggagtaggcccattaatagtatcttt\",\n",
    "    \"test30\": \"ccattaatagcatacaaaatcgactcaag\",\n",
    "    \"test31\": \"ccaattatagaaagctgtggctggtcgtc\",\n",
    "    \"test32\": \"aactattatttctcacattccattaatgg\",\n",
    "    \"test33\": \"atgctttaccaataatagagcgcaa\",\n",
    "    \"test34\": \"ggtcagttagatccaattatggaatg\",\n",
    "    \"test35\": \"gcattccaaaattagtaacgatatct\",\n",
    "    \"test36\": \"cctcctttccaaaattagttgagaag\",\n",
    "    \"test37\": \"ctttgccaaaattagctattctgac\",\n",
    "    \"test38\": \"acgcatgcaccacatatagtaacgtg\",\n",
    "    \"test39\": \"agcgccccatttttagggtttaagct\",\n",
    "    \"test40\": \"ccatttttagtaatttaatacaacgccgc\",\n",
    "    \"test41\": \"ccatttttagtatggaaccgccgtgagt\",\n",
    "    \"test42\": \"aaattaccataagtggtaatgcacacac\",\n",
    "    \"test43\": \"caattactatatatagcgtgtttgtc\",\n",
    "    \"test44\": \"ccatacatggtaaaatgtaccgaaacact\",\n",
    "    \"test45\": \"ccgaaagtactataaatagtaatcca\",\n",
    "    \"test46\": \"gttttcgcctttctataaatagtacc\",\n",
    "    \"test47\": \"ggtacctaatatagtaatcagctctg\",\n",
    "    \"test48\": \"tctttaattacttgcctaatatagct\",\n",
    "    \"test49\": \"ggatgcatccctaatatagttaataa\",\n",
    "    \"test50\": \"atctttaccaaaactagttaattcga\",\n",
    "    \"test51\": \"tgctaaatatagaacatctccaaata\",\n",
    "    \"test52\": \"ccttctaactaaatatagaagtgata\",\n",
    "    \"test53\": \"attaccataaacagaaatcagtggat\",\n",
    "    \"test54\": \"cctaaaaatagatcgaatgtgtgctc\",\n",
    "    \"test55\": \"tactaaaaatagatcatgagctacga\",\n",
    "    \"test56\": \"tagtcacttgatttccatacctaaaatgg\",\n",
    "    \"test57\": \"atcaacagtctacaatatcctaaaatgg\",\n",
    "    \"test58\": \"caaacattccagatttagaatggtta\",\n",
    "    \"test59\": \"acttctttccttttatagctgagtgc\",\n",
    "    \"test60\": \"ccgctaagccttgccttttatagca\",\n",
    "    \"test61\": \"tgcccaaaaaaagaaagttgtcagac\",\n",
    "    \"test62\": \"cccaaataaggaaagctctctggac\",\n",
    "    \"test63\": \"aaccacacacccagaattagtaag\",\n",
    "    \"test64\": \"tggttcatttaagactttaccatttctgg\",\n",
    "    \"test65\": \"tgcgatggaatagtactaaatttagg\",\n",
    "    \"test66\": \"tgcggaagagggttcccgatatagata\",\n",
    "    \"test67\": \"tcccaaaaatactcatatgtcgggct\",\n",
    "    \"test68\": \"ccatattgggtaaagattgctttttagca\",\n",
    "    \"test69\": \"tagtggttaactacccgtttttagta\",\n",
    "    \"test70\": \"ccaatttatagttcactttcgtgatgagaa\",\n",
    "    \"test71\": \"tactcgcttttcttactaaaagtaga\",\n",
    "    \"test72\": \"tactaaaagtagtagtttgtctgca\",\n",
    "    \"test73\": \"ccattttaaggaatttacgatctagtgaa\",\n",
    "    \"test74\": \"tcccattttaagaccaacttctcatt\",\n",
    "    \"test75\": \"gatccagtagattccatttatgtacg\",\n",
    "    \"test76\": \"cctgtaaatagtaacaaggtgcatcg\",\n",
    "    \"test77\": \"cgatcaatatgttaccattttggggt\",\n",
    "    \"test78\": \"gagcgagaccttcccaataattagtaa\",\n",
    "    \"test79\": \"aagtgactatattactcaaaatagaa\",\n",
    "    \"test80\": \"ttaccaaattgccaatttaggctaaa\",\n",
    "    \"test81\": \"ttaccaaattgccaatttaggctaaa\",\n",
    "    \"test82\": \"gaagatacctaatacggaaattttcc\",\n",
    "    \"test83\": \"gcagacttgctatattaagctaatat\",\n",
    "    \"test84\": \"cgctttcttactataaatgtttacta\",\n",
    "    \"test85\": \"ggttcaagatttttccttttatgtac\",\n",
    "    \"test86\": \"tttatagtgtccctttttcggtaagt\",\n",
    "    \"test87\": \"atcctacgatgctttctaaaagaagg\",\n",
    "    \"test88\": \"ccctagcaattttttactatatttgt\",\n",
    "    \"test89\": \"ggaagaatttactataattgtacatg\",\n",
    "    \"test90\": \"ctctaatggccttactacttaaagca\",\n",
    "    \"test91\": \"gccgagtccggaaaatttccgaaaaatg\",\n",
    "    \"test92\": \"atttatttcctaaagtgtaactaac\",\n",
    "    \"test93\": \"gacctgtatccttttctacttttgtc\",\n",
    "    \"test94\": \"atatgcccctcacaagttaccaatta\",\n",
    "    \"test95\": \"aaaatggtaatttctcgggacagg\",\n",
    "    \"test96\": \"ctcagtgcacacagacattccaaata\",\n",
    "    \"test97\": \"gattaggattcgtttgtttccaaata\",\n",
    "}\n",
    "\n",
    "# Common motif analysis to find any recurring subsequences.\n",
    "from Bio import motifs\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "# Convert sequences into Seq objects for motif analysis.\n",
    "seqs = [Seq(sequence) for sequence in sequences.values()]\n",
    "m = motifs.create(seqs)\n",
    "motifs_found = m.consensus\n",
    "\n",
    "motifs_found"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
